tests/fuzz/s2n_select_server_cert_test.c

```diff
@@ -0,0 +1,176 @@
+        }
+
+        /* We found a null byte. Move the cursor beyond it. */
+        cursor = (((const uint8_t *) next_null - buf) + 1);
+        if (cursor >= len) {
+            return num_strings;
+        }
+        output_strings[num_strings] = cur_str;
+        num_strings++;
+    }
+}
+
+GENERAL_NAME *string_to_general_name(const char *str)
+{
+    ASN1_IA5STRING *asn1_name_str = ASN1_IA5STRING_new();
+    if (!asn1_name_str) {
+        return NULL;
+    }
+
+    if (!ASN1_STRING_set(asn1_name_str, str, strlen(str))) {
+        ASN1_IA5STRING_free(asn1_name_str);
+        return NULL;
+    }
+
+    GENERAL_NAME *san_name = GENERAL_NAME_new();
+    if (!san_name) {
+        ASN1_IA5STRING_free(asn1_name_str);
+        return NULL;
+    }
+
+    GENERAL_NAME_set0_value(san_name, GEN_DNS, asn1_name_str);
+    return san_name;
+}
+
+/*
+ * This fuzz test uses the fuzz input to:
+ * - Generate the data to populate in the SAN of a certificate
+ * - Generate the data to populate the SNI TLS extension
+ * - Fuzz the certificate matching function in s2n: s2n_cert_chain_and_key_matches_name
+ */
+int LLVMFuzzerTestOneInput(const uint8_t *buf, size_t len)
+{
+    struct s2n_cert_chain_and_key *cert = s2n_cert_chain_and_key_new();
+    GENERAL_NAMES* san_names = sk_GENERAL_NAME_new_null();
+    X509 *x509_cert = X509_new();
+
+    if (!x509_cert || !san_names || !cert) {
+        goto cleanup;
+    }
+
+    /* Create an array of strings based on fuzz input. To populate the cert SANs and
+     * input hostname.
+     */
+    const char *strings[MAX_TOKENS] = { NULL };
+    size_t num_strings = find_strings(buf, len, strings, MAX_TOKENS);
+    size_t num_sans;
+    /* This should be a match since we make the hostname the same string as a SAN we've added.
+     * It won't be a match when the SAN is longer than domain name length for SNI.
+     */
+    const char *likely_matching_hostname;
+    /* A string that was not part of the set of strings we added for SANs. */
+    const char *other_hostname = NULL;
+    if (num_strings == 0) {
+        likely_matching_hostname = "";
+        num_sans = 0;
+    } else {
+        num_sans = num_strings - 1;
+        likely_matching_hostname = strings[0];
+        other_hostname = strings[num_strings - 1];
+    }
+
+    for (int i = 0; i < num_sans; i++) {
+        GENERAL_NAME *san_name = string_to_general_name(strings[i]);
+        if (!san_name) {
+            continue;
+        }
+        sk_GENERAL_NAME_push(san_names, san_name);
+    }
+
+    if (X509_add1_ext_i2d(x509_cert, NID_subject_alt_name, san_names, 0, X509V3_ADD_REPLACE) <= 0) {
+        goto cleanup;
+    }
+
+    cert->x509_cert = x509_cert;
+    cert->san_names = san_names;
+
+    /* Not checking the return value as we aren't using this fuzz test for matching correctness. */
+    s2n_cert_chain_and_key_matches_name(cert, likely_matching_hostname);
+    if (other_hostname != NULL) {
+        s2n_cert_chain_and_key_matches_name(cert, other_hostname);
```

Comment:
Is this branch just adding additional random coverage? Since you're not testing correctness, is this necessary?

Reply:
It's necessary to try a hostname that isn't one of the SANs. other_hostname should come from a portion of the fuzz input that we didn't load into the cert. I did this in a more clear way in https://github.com/awslabs/s2n/pull/1043 where we just try all strings in the fuzz buffer rather that selecting specific ones.